Sources/PenguinStructures/Tuple.swift

```diff
@@ -0,0 +1,144 @@
+///
+/// Swift's built-in tuple types are algebraic product types, but since they are
+/// not nominal and not easily decomposed, they don't lend themselves to many
+/// types of useful processing.  Models of `Tuple` don't have those problems.
+public protocol AlgebraicProduct {
+  /// The type of the first element.
+  associatedtype Head
+
+  /// An algebriac product formed by composing the remaining elements.
+  associatedtype Tail: AlgebraicProduct
+
+  /// The first element.
+  var head: Head { get set }
+  
+  /// All elements but the first.
+  var tail: Tail { get set }
+}
+
+extension Empty: AlgebraicProduct {
+  /// The first element, when `self` is viewed as an instance of algebraic
+  /// product type.
+  public var head: Never { get {  fatalError() } set {  } }
+  
+  /// All elements but the first, when `self` is viewed as an instance of
+  /// algebraic product type.
+  public var tail: Self { get { self } set { } }
+}
+
+/// An algebraic product type whose first element is of type `Head` and
+/// whose remaining elements can be stored in `Tail`.
+public struct Tuple<Head, Tail: AlgebraicProduct>: AlgebraicProduct {
+  /// The first element.
+  public var head: Head
+  
+  /// All elements but the first.
+  public var tail: Tail
+}
+
+extension Tuple: DefaultInitializable
+  where Head: DefaultInitializable, Tail: DefaultInitializable
+{
+  // Initialize `self`.
+  public init() {
+    head = Head()
+    tail = Tail()
+  }
+}
+
+extension Tuple: Equatable where Head: Equatable, Tail: Equatable {}
+extension Tuple: Hashable where Head: Hashable, Tail: Hashable {}
+extension Tuple: Comparable where Head: Comparable, Tail: Comparable {
+  public static func < (lhs: Self, rhs: Self) -> Bool {
+    if lhs.head < rhs.head { return true }
+    if lhs.head > rhs.head { return false }
+    return lhs.tail < rhs.tail
+  }
+}
+
+private let prefixLength = "Tuple(".count
+
+extension Tuple: CustomStringConvertible {
+  public var description: String {
+    if Tail.self == Empty.self {
+      return "Tuple(\(String(reflecting:head )))"
+    }
+    else {
+      return "Tuple(\(String(reflecting: head)), "
+        + String(reflecting: tail).dropFirst(prefixLength)
+    }
+  }
+}
+
+// ======== Conveniences ============
+
+public typealias Tuple0 = Empty
+public typealias Tuple1<T0> = Tuple<T0, Tuple0>
+public typealias Tuple2<T0, T1> = Tuple<T0, Tuple1<T1>>
+public typealias Tuple3<T0, T1, T2> = Tuple<T0, Tuple2<T1, T2>>
+public typealias Tuple4<T0, T1, T2, T3> = Tuple<T0, Tuple3<T1, T2, T3>>
+public typealias Tuple5<T0, T1, T2, T3, T4> = Tuple<T0, Tuple4<T1, T2, T3, T4>>
+public typealias Tuple6<T0, T1, T2, T3, T4, T5>
+  = Tuple<T0, Tuple5<T1, T2, T3, T4, T5>>
+public typealias Tuple7<T0, T1, T2, T3, T4, T5, T6>
+  = Tuple<T0, Tuple6<T1, T2, T3, T4, T5, T6>>
+
+public extension Tuple {
+  /// Creates an instance containing the arguments, in order.
+  init(_ a0: Head) where Tail == Empty {
+    head = a0; tail = Tuple0()
+  }  
```

Comment:
I think you can get this to compile by writing:

```suggestion
public extension Tuple where Tail == Empty {
  /// Creates an instance containing the arguments, in order.
  init(_ a0: Head) {
    head = a0; tail = Tuple0()
  }
}
public extension Tuple {
```

Reply:
Could swear I did compile it.  Heh, works with our Swift but not Apple's.  Thanks!